### DISCUSSION TITLE
Experimentation with the NRF PPKII, current measurements

### DISCUSSION THREAD

**Discussion by JF002** [InfiniTimeOrg/InfiniTime#1420] [Show and tell]
A few weeks ago, I finally received my brand new [NRF PPKII](https://www.nordicsemi.com/Products/Development-hardware/Power-Profiler-Kit-2). It's a relatively cheap board that allows to easily measure and monitor the power usage (current) of any device.

I've finally found some time to hook it up on a NRF52DK (development board for the NRF52832) and on my PineTime devkit. Here are some results.

First, how low can the NRF52832 can go? According to the datasheet, it should use 1.2µA in SYSTEM ON mode (the mode implemented in the InfiniTime) and 0.7µA in SYSTEM OFF mode. In this test, I wrote a very simple code that simply puts the MCU in SYSTEM OFF mode after 1 second of runtime. I measured 25µA:
![image](https://user-images.githubusercontent.com/2261652/200172944-1a486a2c-b3e0-46af-a0a1-7581994545bb.png)

The very same code on the PineTime shows much higher results : 1.2mA
![image](https://user-images.githubusercontent.com/2261652/200173315-51e756db-d7b7-40a0-a2a6-dd6ed5b02e9a.png)

This increase in current measurement can probably be explain by all the other components and chips (SPI flash, battery charger, sensors,...) that are powered by the PineTime battery.

Finally I measured how much current InfiniTime actually uses:
![image](https://user-images.githubusercontent.com/2261652/200173334-4ef1fef5-43c4-4f9c-a870-705f8ae2fc4e.png)

It looks like InfiniTime does a really good job with power management : the difference between a simple code that puts the MCU in SYSTEMOFF mode and the whole InfiniTime app that stays in SYSTEM ON mode (deep sleep) is very thin. It even seems to use a bit less current, probably because it puts all external devices to sleep before going to deep sleep.

**Comment by Riksu9000** [comment]
I still think there must be something we're missing. This person was looking into it and got the usage a bit lower than here. https://github.com/InfiniTimeOrg/InfiniTime/issues/53#issuecomment-791725224 https://github.com/InfiniTimeOrg/InfiniTime/issues/53#issuecomment-782417587

Here he claims to get just 66uA. I'm not 100% sure which hardware this was on though. https://github.com/InfiniTimeOrg/InfiniTime/issues/53#issuecomment-783654321

**Reply by JF002** [reply]
Yes, you're right... I haven't been able to go below ~1mA so far... 66µA in sleep mode would certainly improve the battery life by A LOT!

**Reply by aykevl** [reply]
I was able to reach ~0.18mA on the 3.3V pin (battery disconnected). I have posted my findings on the wiki: https://wiki.pine64.org/wiki/PineTime#Reducing_power_consumption
I'm not entirely sure about those 66µA. It would certainly be great, but I don't see many ways to reduce current consumption even further. In fact I tried reproducing the setup with Mynewt and got a very similar result as before: 0.18mA.

(There is of course the possibility that my measurements are wrong too, I intend to also get a Power Profiler Kit to be sure).

Update: bought a power profiler kit, it'll arrive in a few days!

**Reply by aykevl** [reply]
Managed to reproduce ~60µA, so yes it does appear possible! I'll get my own Power Profiler Kit very soon and I'll test again, in case my multimeter is wrong (for example, there might be short current spikes that it doesn't pick up).
See my comment here: https://github.com/InfiniTimeOrg/InfiniTime/issues/53#issuecomment-1517032936

Update: the Power Profiler Kit measures around 72µA current on average, see the issue for details.

**Comment by JF002** [comment]
Following @Riksu9000's idea, I measured the current used by the touchpanel, and tried disabling a few features from the touchpanel to check if it has any impact on power consumption.

Here's an overview of the power consumption when the InfiniTime is in sleep mode:
![power-touchpanel1](https://user-images.githubusercontent.com/2261652/210398451-6429a4e0-b7aa-4a9b-bb1f-24001c8e4ac9.png)

As you can see, touching the display increases the current from 1.1mA to 3.6 during 2s, even if all wake options are disabled. I guess this is the touchpanel going out of sleep mode as soon as it detects a touch.

I tried disabling EnConLR, EnTouch and EnChange but I couldn't see any change in the current measurement.

![image](https://user-images.githubusercontent.com/2261652/210399070-c418e4e8-290e-4630-ac44-9490b055429e.png)

**Comment by JF002** [comment]
Following @aykevl suggestions, I opened a few PR:
 - #1755
 - #1756 
 - #1757 
 - #1758 

I merged all of them except #1757 which still requires a bit of work.
I think we should be able to reduce the power usage in sleep mode by disabling a few additional GPIO (SPI CS and the display reset pin), but I'm not sure if it's safe to do this. 

Here's a comparison of the power usage of InfiniTime 1.12 and InfiniTime with those changes:

---------------------
| Version | BLE Fast Advertising | BLE Slow Advertising | BLE connected | BLE OFF |
|-|-|-|-|-|
| InfiniTime 1.12 | 1.5mA | 890µA | 1.3mA | 860µA
| Current dev branch with power opimizations | 820µA | 280µA | 380µA | 250µA
----------------------

When I tested those changes, my PineTime ran for a bit more than 16 days on a single charge!
